apps/safety/main.cpp

```diff
@@ -1,13 +1,219 @@
+CANMsg rxMsg;
+//needs a specific CANID to update values
+const unsigned int expectedCANID = 0x000;
+
+//  I2C bus initialization
+I2C i2c(CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL); //pinnames found from pinnames.h
+const int SENSOR_SLAVE_ADDRESS = 0x1c << 1;
+
+enum configurationMode{
+    powerReset = 0,
+    triggered = 3,
+    continous = 7
+};
 
-const int k_interval_ms = 500;
+//below are settings for the INA226 chip. The default values have been selected 
+//but different conversion times can be selected as well. See datasheet for other settings
+const int busVoltageConversionTime = 4;
+const int shuntVoltageConversionTime = 4;
+const int averageMode = 0;
+
+// temperature conversion constants
+const float A = 0;
+const float B = 0;
+const float C = 0;
+const float rResistor = 0;
+
+//measurement thresholds, needs updating
+float currentUpperLim = 1;
+float currentLowerLim = 0;
+float voltageUpperLim = 1;
+float voltageLowerLim = 0;
+float tempUpperLim = 1;
+float tempLowerLim = 0;
+
+//funcs
+int inRange(float upperLim, float lowerLim, float val);
+float getCurrentData();
+float getVoltageData();
+float readThermosistor();
+int calibrateSensor();
+int configureSensor(configurationMode operationMode, int resetRegisters);
+void initCAN();
+void processCANMsg(CANMsg *p_newMsg);
 
 int main()
 {
+    // setup bus
+    i2c.frequency(10000);
+    initCAN();
+    // turn led off, turn on if issue occurs
+    led1 = 0;
+    led2 = 0;
+    led3 = 0;
+
     while (true) {
-        led1 = !led1;
-        wait_ms(k_interval_ms);
+        //read can messages to get data to configure device
+        u_int8_t command = 0;
+        if(can.read(rxMsg))
+        {
+            processCANMsg(&rxMsg);
+            rxMsg >> command;
+            rxMsg.clear();
+        }
+
+        float measuredCurrent = getCurrentData();
+        if (inRange(currentUpperLim, currentLowerLim, measuredCurrent))
+        {
+            led1 = 1;
+        }
+
+        float measuredVoltage = getVoltageData();
+        if (inRange(voltageUpperLim, voltageLowerLim, measuredVoltage))
+        {
+            led2 = 1;
+        }
+
+        float measuredTemp = readThermosistor();
+        if(inRange(tempUpperLim, tempLowerLim, measuredTemp))
+        {
+            led3 = 1;
+        }
     }
 }
+
+void initCAN()
+{
+    // will need new id's 
+    can.filter(CAN1_RX, CAN1_TX, ROVER_CANID_MASK, CANStandard);
+}
+
+void processCANMsg(CANMsg *p_newMsg)
+{
+    // PRINT_INFO("Recieved CAN message with ID %X\r\n", p_newMsg->id);
+    // The specific can ID for configuring chip is ___
+    switch (p_newMsg->id){
+        case expectedCANID:
+            pc.printf("Updating INA226 config\r\n");
+            break;
+        default:
+            pc.printf("Recieved unimplemented command\r\n");
+            break;
+    }
+}
+
+
+int inRange(float upperLim, float lowerLim, float val)
+{
+    int inRange = (val > upperLim || val< lowerLim) ? 1: 0;
```

Comment:
Logic is wrong, should be (val > upperLim || val< lowerLim) ? 0 : 1; or  (val < upperLim && val > lowerLim) ? 1 : 0